2024/incoming/have-an-ai-christmas.pod

```diff
@@ -0,0 +1,282 @@
+
+But who has time for that now? Nowadays, if my family sends out cards
+at all, we buy them in bulk at the local stationery store. Santa would
+not approve!
+
+=head2 AI to the Rescue!
+
+We've already surrendered our coding to Copilot, our research to
+ChatGPT, and some of us are replacing our real-life friends with
+simulated personalities on Character.ai. So why not let AI create our
+Christmas greetings for us?
+
+L<OpenAI::API> to the rescue! This handy-dandy Perl module exposes
+OpenAI's API to Perl, and lets you do chat, text completion, and
+image generation from the command line. What we're going to do is:
+
+=over
+
+=item 1. Get a description of the type of card to generate and (optionally) the name of its recipient.
+
+=item 2. Call OpenAI's chat API to turn the description into an image generation prompt.
+
+=item 3. Call OpenAI's image generation API to generate an image from the prompt and store it online.
+
+=item 4. Use L<Browser::Open> to display the image in your favorite web browser.
+
+=back
+
+When you run the script it does this:
+
+   > perl merry_xmas.pl "A humorous card for my Uncle Jim, who loves cats"
+   The AI elves are hard at work (tippity tappity)...done!
+   
+Your browser will then open up and display something like this:
+
+=for html
+<img src="xmas-card-for-uncle-jim.png" />
+
+=head2 Prequisites
+
+For this script to work, you'll need to install the L<Browser::Open>
+and L<OpenAI::API> modules from CPAN. You'll also need to register an
+OpenAI account, get an API key, and add the key to your environment
+using the name L<OPENAI_API_KEY>. Instructions for this can be found
+at L<https://platform.openai.com/docs/quickstart>.
+
+=head2 The Main Script!
+
+Here's the main script. Save it as merry_xmas.pl:
+
+=begin perl
+
+#!/usr/bin/env perl
+use lib '.','./lib','../lib';
+use strict;
+use Browser::Open 'open_browser';
+use ChristmasCard;
+
+die <<END unless @ARGV;
+Usage: $0 "card description"
+   Generate a Christmas card from a brief description.
+   If the description contains the recipient's name, the name will
+   be incorporated in the card's greeting.
+END
+
+my $description = "@ARGV";
+
+print STDERR "The AI elves are hard at work (tippity tappity)...";
+my $url = ChristmasCard->make_card($description);
+print STDERR "...done!\n";
+
+open_browser($url);
+
+=end perl
+
+This script imports the C<ChristmasCard.pm> module which does the
+actual work. It then uses C<@ARGV> to read the description of the card
+from the command line and passes it to the
+C<ChristmasCard->>make_card()> method. This method returns a URL to the
+image and passes it to the L<Browser::Open> C<open_browser()> function
+to display the image. You can then download the image, attach it to an
+email, etc.
+
+=head2 The ChristmasCard library
+
+The second and last piece is the C<ChristmasCard> library. Save this
+file as C<ChristmasCard.pm> either in the same directory as the main
+script, or in a subdirectory named C<lib>.
```

Comment:
I think if we changed this guidance to "Save this in ./lib/ChristmasCard.pm", then we could simplify the `use lib` statement in the code sample above.

Reply:
Thanks. I fixed this up to use the simpler "use lib".